mungegithub/mungers/submit-queue-batch.go

```diff
@@ -0,0 +1,282 @@
+	BaseName string
+	BaseSha  string
+	Pulls    []BatchPull
+}
+
+// batchRefToBatch parses a string into a Batch
+func batchRefToBatch(batchRef string) (Batch, error) {
+	batch := Batch{}
+	for i, ref := range strings.Split(batchRef, ",") {
+		parts := strings.Split(ref, ":")
+		if len(parts) != 2 {
+			return Batch{}, errors.New("bad batchref")
+		}
+		if i == 0 {
+			batch.BaseName = parts[0]
+			batch.BaseSha = parts[1]
+		} else {
+			num, err := strconv.ParseInt(parts[0], 10, 32)
+			if err != nil {
+				return Batch{}, err
+			}
+			batch.Pulls = append(batch.Pulls, BatchPull{int(num), parts[1]})
+		}
+	}
+	return batch, nil
+}
+
+// getCompleteBatches returns a list of Batches that passed all
+// required tests.
+func (sq *SubmitQueue) getCompleteBatches(jobs []ProwJob) []Batch {
+	batchContexts := make(map[string]map[string]interface{})
+	for _, job := range jobs {
+		if _, ok := batchContexts[job.Refs]; !ok {
+			batchContexts[job.Refs] = make(map[string]interface{})
+		}
+		batchContexts[job.Refs][job.Context] = nil
+	}
+	batches := []Batch{}
+	for batchRef, contexts := range batchContexts {
+		match := true
+		// Did this succeed in all the contexts we want?
+		for _, ctx := range sq.RequiredStatusContexts {
+			if _, ok := contexts[ctx]; !ok {
+				match = false
+			}
+		}
+		for _, ctx := range sq.RequiredRetestContexts {
+			if _, ok := contexts[ctx]; !ok {
+				match = false
+			}
+		}
+		if match {
+			batch, err := batchRefToBatch(batchRef)
+			if err != nil {
+				continue
+			}
+			batches = append(batches, batch)
+		}
+	}
+	return batches
+}
+
+// batchIntersectsQueue returns whether at least one PR in the batch is queued.
+func (sq *SubmitQueue) batchIntersectsQueue(batch Batch) bool {
+	sq.Lock()
+	defer sq.Unlock()
+	for _, pull := range batch.Pulls {
+		if _, ok := sq.githubE2EQueue[pull.Number]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+// matchesCommit determines if the batch can be merged given some commits.
+// That is, does it contain exactly:
+// 1) the batch's BaseSha
+// 2) (optional) merge commits for PRs in the batch
+// 3) any merged PRs in the batch are sequential from the beginning
+// The return value is -1 on error, else the number of PRs already merged.
+func (batch *Batch) matchesCommits(commits []*githubapi.RepositoryCommit) int {
+	if len(commits) == 0 {
+		return -1
+	}
+
+	shaToPR := make(map[string]int)
+
+	for _, pull := range batch.Pulls {
+		shaToPR[pull.Sha] = pull.Number
+	}
+
+	matchedPRs := []int{}
+
+	// convert the list of commits into a DAG for easy following
+	dag := make(map[string]*githubapi.RepositoryCommit)
+	for _, commit := range commits {
+		dag[*commit.SHA] = commit
+	}
+
+	ref := *commits[0].SHA
+	for {
+		if ref == batch.BaseSha {
+			break // found the base ref (condition #1)
+		}
+		commit, ok := dag[ref]
+		if !ok {
+			return -1 // ref not in partial list of commits
+		}
+		if len(commit.Parents) == 2 && commit.Message != nil && (*commit.Message)[:5] == "Merge" {
```

Comment:
`(*commit.Message)[:5] == "Merge"` :-/

Do we really need that? Also, if we do, it should probably be `strings.HasPrefix(*commit.Message, "Merge")`.